### ISSUE TITLE
[ModuleClient] fails to start when calling createFromEnvironment

### ISSUE THREAD

**Issue by hochas** [Azure/azure-iot-sdk-java#434]
- **OS and version used:** 
Linux raspberrypi 4.14.79-v7+ #1159 SMP Sun Nov 4 17:50:20 GMT 2018 armv7l GNU/Linux

- **Java runtime used:**
openjdk version "1.8.0_171"
OpenJDK Runtime Environment (build 1.8.0_171-8u171-b11-1~deb9u1-b11)
OpenJDK Client VM (build 25.171-b11, mixed mode)

- **SDK version used:** 1.14.2

- **IoT Edge runtime:** iotedge 1.0.6 (8288bc9bd6f6e15295fea506cd3f99d7f6347a6a)

# Description of the issue:

We are deplyoing an edge module using spring boot to a raspberry pi.
We are launching our application in two different ways depending on environment:

``` 
private ModuleClient createModule(IotHubClientProtocol protocol) throws ModuleClientException, URISyntaxException {
	if (createFromEnvironment)
		return ModuleClient.createFromEnvironment(protocol);
	else {
		return new ModuleClient("connectionString", protocol);
	}
}
``` 

Calling the static constructor method always fails when we deploy through Azure Portal (is there a faster way of launching a module using the iot edge runtime than going through the portal and using *Set modules*?). Creating the ModuleClient explicitly with `new` always works, both in docker and in the iotedge runtime. We have the exact same connection string for both scenarios. Nothing is changed in `/etc/iotedge/config.yaml` apart from the manual provisioning details.

If I add tomcat as a dependency:

```
<dependency>
	<groupId>org.springframework.boot</groupId>
	<artifactId>spring-boot-starter-web</artifactId>
</dependency>
```

I get the exact same error as #274.

Is there any way I can debug my application while in the context of the iotedge runtime, as this error only occurs when that is in use?

**Additional details:**

I added the following to my Dockerfile:

```
ENV EdgeHubConnectionString connectionString
```
And with that it works to call `createFromEnvironment` both from docker on my dev machine and when deploying to the iotedge runtime from Azure portal.

# Code sample exhibiting the issue:

# Console log of the issue:
```
[INFO] Application - Starting Application v0.0.1-SNAPSHOT on 214221401e82 with PID 1 (/app/app.jar started by moduleuser in /app)
[INFO] Application - No active profile set, falling back to default profiles: default
[INFO] ThreadPoolTaskScheduler - Initializing ExecutorService 'defaultTaskScheduler'
[INFO] Application - Started Application in 8.274 seconds (JVM running for 10.108)
Exception in thread "main" java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:48)
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:87)
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:50)
	at org.springframework.boot.loader.JarLauncher.main(JarLauncher.java:51)
Caused by: java.lang.UnsatisfiedLinkError: could not load FFI provider jnr.ffi.provider.jffi.Provider
	at jnr.ffi.provider.InvalidRuntime.newLoadError(InvalidRuntime.java:101)
	at jnr.ffi.provider.InvalidRuntime.findType(InvalidRuntime.java:42)
	at jnr.ffi.Struct$NumberField.<init>(Struct.java:872)
	at jnr.ffi.Struct$Unsigned16.<init>(Struct.java:1240)
	at jnr.unixsocket.SockAddrUnix$DefaultSockAddrUnix.<init>(SockAddrUnix.java:209)
	at jnr.unixsocket.SockAddrUnix.create(SockAddrUnix.java:174)
	at jnr.unixsocket.UnixSocketAddress.<init>(UnixSocketAddress.java:53)
	at com.microsoft.azure.sdk.iot.device.hsm.HttpsHsmClient.sendHttpRequestUsingUnixSocket(HttpsHsmClient.java:250)
	at com.microsoft.azure.sdk.iot.device.hsm.HttpsHsmClient.sendRequestBasedOnScheme(HttpsHsmClient.java:226)
	at com.microsoft.azure.sdk.iot.device.hsm.HttpsHsmClient.sign(HttpsHsmClient.java:112)
	at com.microsoft.azure.sdk.iot.device.hsm.HttpHsmSignatureProvider.sign(HttpHsmSignatureProvider.java:85)
	at com.microsoft.azure.sdk.iot.device.hsm.IotHubSasTokenHsmAuthenticationProvider.createNewSasToken(IotHubSasTokenHsmAuthenticationProvider.java:80)
	at com.microsoft.azure.sdk.iot.device.hsm.IotHubSasTokenHsmAuthenticationProvider.create(IotHubSasTokenHsmAuthenticationProvider.java:48)
	at com.microsoft.azure.sdk.iot.device.ModuleClient.createFromEnvironment(ModuleClient.java:235)
	at storemanager.azure.EdgeModuleHandler.createModule(EdgeModuleHandler.java:81)
	at storemanager.azure.EdgeModuleHandler.startModule(EdgeModuleHandler.java:73)
	at storemanager.azure.EdgeModuleHandler.start(EdgeModuleHandler.java:48)
	at storemanager.Application.main(Application.java:14)
	... 8 more
Caused by: java.lang.ExceptionInInitializerError
	at jnr.ffi.provider.jffi.NativeRuntime.getInstance(NativeRuntime.java:58)
	at jnr.ffi.provider.jffi.Provider.<init>(Provider.java:29)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at java.lang.Class.newInstance(Class.java:442)
	at jnr.ffi.provider.FFIProvider$SystemProviderSingletonHolder.getInstance(FFIProvider.java:68)
	at jnr.ffi.provider.FFIProvider$SystemProviderSingletonHolder.<clinit>(FFIProvider.java:57)
	at jnr.ffi.provider.FFIProvider.getSystemProvider(FFIProvider.java:35)
	at jnr.ffi.Runtime$SingletonHolder.<clinit>(Runtime.java:82)
	at jnr.ffi.Runtime.getSystemRuntime(Runtime.java:67)
	at jnr.unixsocket.SockAddrUnix.<init>(SockAddrUnix.java:46)
	at jnr.unixsocket.SockAddrUnix$DefaultSockAddrUnix.<init>(SockAddrUnix.java:208)
	... 21 more
Caused by: java.lang.IllegalStateException: Can't overwrite cause with java.lang.UnsatisfiedLinkError: java.lang.UnsatisfiedLinkError: /tmp/jffi726799928824727829.so: /tmp/jffi726799928824727829.so: cannot open shared object file: No such file or directory
	at java.lang.ClassLoader$NativeLibrary.load(Native Method)
	at java.lang.ClassLoader.loadLibrary0(ClassLoader.java:1941)
	at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1824)
	at java.lang.Runtime.load0(Runtime.java:809)
	at java.lang.System.load(System.java:1086)
	at com.kenai.jffi.internal.StubLoader.loadFromJar(StubLoader.java:371)
	at com.kenai.jffi.internal.StubLoader.load(StubLoader.java:258)
	at com.kenai.jffi.internal.StubLoader.<clinit>(StubLoader.java:444)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:348)
	at com.kenai.jffi.Init.load(Init.java:68)
	at com.kenai.jffi.Foreign$InstanceHolder.getInstanceHolder(Foreign.java:49)
	at com.kenai.jffi.Foreign$InstanceHolder.<clinit>(Foreign.java:45)
	at com.kenai.jffi.Foreign.getInstance(Foreign.java:103)
	at com.kenai.jffi.Type$Builtin.lookupTypeInfo(Type.java:242)
	at com.kenai.jffi.Type$Builtin.getTypeInfo(Type.java:237)
	at com.kenai.jffi.Type.resolveSize(Type.java:155)
	at com.kenai.jffi.Type.size(Type.java:138)
	at jnr.ffi.provider.jffi.NativeRuntime$TypeDelegate.size(NativeRuntime.java:187)
	at jnr.ffi.provider.AbstractRuntime.<init>(AbstractRuntime.java:48)
	at jnr.ffi.provider.jffi.NativeRuntime.<init>(NativeRuntime.java:66)
	at jnr.ffi.provider.jffi.NativeRuntime.<init>(NativeRuntime.java:41)
	at jnr.ffi.provider.jffi.NativeRuntime$SingletonHolder.<clinit>(NativeRuntime.java:62)
	at jnr.ffi.provider.jffi.NativeRuntime.getInstance(NativeRuntime.java:58)
	at jnr.ffi.provider.jffi.Provider.<init>(Provider.java:29)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at java.lang.Class.newInstance(Class.java:442)
	at jnr.ffi.provider.FFIProvider$SystemProviderSingletonHolder.getInstance(FFIProvider.java:68)
	at jnr.ffi.provider.FFIProvider$SystemProviderSingletonHolder.<clinit>(FFIProvider.java:57)
	at jnr.ffi.provider.FFIProvider.getSystemProvider(FFIProvider.java:35)
	at jnr.ffi.Runtime$SingletonHolder.<clinit>(Runtime.java:82)
	at jnr.ffi.Runtime.getSystemRuntime(Runtime.java:67)
	at jnr.unixsocket.SockAddrUnix.<init>(SockAddrUnix.java:46)
	at jnr.unixsocket.SockAddrUnix$DefaultSockAddrUnix.<init>(SockAddrUnix.java:208)
	at jnr.unixsocket.SockAddrUnix.create(SockAddrUnix.java:174)
	at jnr.unixsocket.UnixSocketAddress.<init>(UnixSocketAddress.java:53)
	at com.microsoft.azure.sdk.iot.device.hsm.HttpsHsmClient.sendHttpRequestUsingUnixSocket(HttpsHsmClient.java:250)
	at com.microsoft.azure.sdk.iot.device.hsm.HttpsHsmClient.sendRequestBasedOnScheme(HttpsHsmClient.java:226)
	at com.microsoft.azure.sdk.iot.device.hsm.HttpsHsmClient.sign(HttpsHsmClient.java:112)
	at com.microsoft.azure.sdk.iot.device.hsm.HttpHsmSignatureProvider.sign(HttpHsmSignatureProvider.java:85)
	at com.microsoft.azure.sdk.iot.device.hsm.IotHubSasTokenHsmAuthenticationProvider.createNewSasToken(IotHubSasTokenHsmAuthenticationProvider.java:80)
	at com.microsoft.azure.sdk.iot.device.hsm.IotHubSasTokenHsmAuthenticationProvider.create(IotHubSasTokenHsmAuthenticationProvider.java:48)
	at com.microsoft.azure.sdk.iot.device.ModuleClient.createFromEnvironment(ModuleClient.java:235)
	at storemanager.azure.EdgeModuleHandler.createModule(EdgeModuleHandler.java:81)
	at storemanager.azure.EdgeModuleHandler.startModule(EdgeModuleHandler.java:73)
	at storemanager.azure.EdgeModuleHandler.start(EdgeModuleHandler.java:48)
	at storemanager.Application.main(Application.java:14)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:48)
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:87)
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:50)
	at org.springframework.boot.loader.JarLauncher.main(JarLauncher.java:51)

	at java.lang.Throwable.initCause(Throwable.java:457)
	at com.kenai.jffi.Type$Builtin.lookupTypeInfo(Type.java:252)
	at com.kenai.jffi.Type$Builtin.getTypeInfo(Type.java:237)
	at com.kenai.jffi.Type.resolveSize(Type.java:155)
	at com.kenai.jffi.Type.size(Type.java:138)
	at jnr.ffi.provider.jffi.NativeRuntime$TypeDelegate.size(NativeRuntime.java:187)
	at jnr.ffi.provider.AbstractRuntime.<init>(AbstractRuntime.java:48)
	at jnr.ffi.provider.jffi.NativeRuntime.<init>(NativeRuntime.java:66)
	at jnr.ffi.provider.jffi.NativeRuntime.<init>(NativeRuntime.java:41)
	at jnr.ffi.provider.jffi.NativeRuntime$SingletonHolder.<clinit>(NativeRuntime.java:62)
	... 35 more
Caused by: java.lang.UnsatisfiedLinkError: java.lang.UnsatisfiedLinkError: /tmp/jffi726799928824727829.so: /tmp/jffi726799928824727829.so: cannot open shared object file: No such file or directory
	at java.lang.ClassLoader$NativeLibrary.load(Native Method)
	at java.lang.ClassLoader.loadLibrary0(ClassLoader.java:1941)
	at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1824)
	at java.lang.Runtime.load0(Runtime.java:809)
	at java.lang.System.load(System.java:1086)
	at com.kenai.jffi.internal.StubLoader.loadFromJar(StubLoader.java:371)
	at com.kenai.jffi.internal.StubLoader.load(StubLoader.java:258)
	at com.kenai.jffi.internal.StubLoader.<clinit>(StubLoader.java:444)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:348)
	at com.kenai.jffi.Init.load(Init.java:68)
	at com.kenai.jffi.Foreign$InstanceHolder.getInstanceHolder(Foreign.java:49)
	at com.kenai.jffi.Foreign$InstanceHolder.<clinit>(Foreign.java:45)
	at com.kenai.jffi.Foreign.getInstance(Foreign.java:103)
	at com.kenai.jffi.Type$Builtin.lookupTypeInfo(Type.java:242)
	at com.kenai.jffi.Type$Builtin.getTypeInfo(Type.java:237)
	at com.kenai.jffi.Type.resolveSize(Type.java:155)
	at com.kenai.jffi.Type.size(Type.java:138)
	at jnr.ffi.provider.jffi.NativeRuntime$TypeDelegate.size(NativeRuntime.java:187)
	at jnr.ffi.provider.AbstractRuntime.<init>(AbstractRuntime.java:48)
	at jnr.ffi.provider.jffi.NativeRuntime.<init>(NativeRuntime.java:66)
	at jnr.ffi.provider.jffi.NativeRuntime.<init>(NativeRuntime.java:41)
	at jnr.ffi.provider.jffi.NativeRuntime$SingletonHolder.<clinit>(NativeRuntime.java:62)
	at jnr.ffi.provider.jffi.NativeRuntime.getInstance(NativeRuntime.java:58)
	at jnr.ffi.provider.jffi.Provider.<init>(Provider.java:29)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at java.lang.Class.newInstance(Class.java:442)
	at jnr.ffi.provider.FFIProvider$SystemProviderSingletonHolder.getInstance(FFIProvider.java:68)
	at jnr.ffi.provider.FFIProvider$SystemProviderSingletonHolder.<clinit>(FFIProvider.java:57)
	at jnr.ffi.provider.FFIProvider.getSystemProvider(FFIProvider.java:35)
	at jnr.ffi.Runtime$SingletonHolder.<clinit>(Runtime.java:82)
	at jnr.ffi.Runtime.getSystemRuntime(Runtime.java:67)
	at jnr.unixsocket.SockAddrUnix.<init>(SockAddrUnix.java:46)
	at jnr.unixsocket.SockAddrUnix$DefaultSockAddrUnix.<init>(SockAddrUnix.java:208)
	at jnr.unixsocket.SockAddrUnix.create(SockAddrUnix.java:174)
	at jnr.unixsocket.UnixSocketAddress.<init>(UnixSocketAddress.java:53)
	at com.microsoft.azure.sdk.iot.device.hsm.HttpsHsmClient.sendHttpRequestUsingUnixSocket(HttpsHsmClient.java:250)
	at com.microsoft.azure.sdk.iot.device.hsm.HttpsHsmClient.sendRequestBasedOnScheme(HttpsHsmClient.java:226)
	at com.microsoft.azure.sdk.iot.device.hsm.HttpsHsmClient.sign(HttpsHsmClient.java:112)
	at com.microsoft.azure.sdk.iot.device.hsm.HttpHsmSignatureProvider.sign(HttpHsmSignatureProvider.java:85)
	at com.microsoft.azure.sdk.iot.device.hsm.IotHubSasTokenHsmAuthenticationProvider.createNewSasToken(IotHubSasTokenHsmAuthenticationProvider.java:80)
	at com.microsoft.azure.sdk.iot.device.hsm.IotHubSasTokenHsmAuthenticationProvider.create(IotHubSasTokenHsmAuthenticationProvider.java:48)
	at com.microsoft.azure.sdk.iot.device.ModuleClient.createFromEnvironment(ModuleClient.java:235)
	at storemanager.azure.EdgeModuleHandler.createModule(EdgeModuleHandler.java:81)
	at storemanager.azure.EdgeModuleHandler.startModule(EdgeModuleHandler.java:73)
	at storemanager.azure.EdgeModuleHandler.start(EdgeModuleHandler.java:48)
	at storemanager.Application.main(Application.java:14)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:48)
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:87)
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:50)
	at org.springframework.boot.loader.JarLauncher.main(JarLauncher.java:51)

	at com.kenai.jffi.Foreign.newLoadError(Foreign.java:72)
	at com.kenai.jffi.Foreign.access$300(Foreign.java:42)
	at com.kenai.jffi.Foreign$InValidInstanceHolder.getForeign(Foreign.java:98)
	at com.kenai.jffi.Foreign.getInstance(Foreign.java:103)
	at com.kenai.jffi.Type$Builtin.lookupTypeInfo(Type.java:242)
	... 43 more
Caused by: java.lang.UnsatisfiedLinkError: java.lang.UnsatisfiedLinkError: /tmp/jffi726799928824727829.so: /tmp/jffi726799928824727829.so: cannot open shared object file: No such file or directory
	at java.lang.ClassLoader$NativeLibrary.load(Native Method)
	at java.lang.ClassLoader.loadLibrary0(ClassLoader.java:1941)
	at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1824)
	at java.lang.Runtime.load0(Runtime.java:809)
	at java.lang.System.load(System.java:1086)
	at com.kenai.jffi.internal.StubLoader.loadFromJar(StubLoader.java:371)
	at com.kenai.jffi.internal.StubLoader.load(StubLoader.java:258)
	at com.kenai.jffi.internal.StubLoader.<clinit>(StubLoader.java:444)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:348)
	at com.kenai.jffi.Init.load(Init.java:68)
	at com.kenai.jffi.Foreign$InstanceHolder.getInstanceHolder(Foreign.java:49)
	at com.kenai.jffi.Foreign$InstanceHolder.<clinit>(Foreign.java:45)
	at com.kenai.jffi.Foreign.getInstance(Foreign.java:103)
	at com.kenai.jffi.Type$Builtin.lookupTypeInfo(Type.java:242)
	at com.kenai.jffi.Type$Builtin.getTypeInfo(Type.java:237)
	at com.kenai.jffi.Type.resolveSize(Type.java:155)
	at com.kenai.jffi.Type.size(Type.java:138)
	at jnr.ffi.provider.jffi.NativeRuntime$TypeDelegate.size(NativeRuntime.java:187)
	at jnr.ffi.provider.AbstractRuntime.<init>(AbstractRuntime.java:48)
	at jnr.ffi.provider.jffi.NativeRuntime.<init>(NativeRuntime.java:66)
	at jnr.ffi.provider.jffi.NativeRuntime.<init>(NativeRuntime.java:41)
	at jnr.ffi.provider.jffi.NativeRuntime$SingletonHolder.<clinit>(NativeRuntime.java:62)
	at jnr.ffi.provider.jffi.NativeRuntime.getInstance(NativeRuntime.java:58)
	at jnr.ffi.provider.jffi.Provider.<init>(Provider.java:29)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at java.lang.Class.newInstance(Class.java:442)
	at jnr.ffi.provider.FFIProvider$SystemProviderSingletonHolder.getInstance(FFIProvider.java:68)
	at jnr.ffi.provider.FFIProvider$SystemProviderSingletonHolder.<clinit>(FFIProvider.java:57)
	at jnr.ffi.provider.FFIProvider.getSystemProvider(FFIProvider.java:35)
	at jnr.ffi.Runtime$SingletonHolder.<clinit>(Runtime.java:82)
	at jnr.ffi.Runtime.getSystemRuntime(Runtime.java:67)
	at jnr.unixsocket.SockAddrUnix.<init>(SockAddrUnix.java:46)
	at jnr.unixsocket.SockAddrUnix$DefaultSockAddrUnix.<init>(SockAddrUnix.java:208)
	at jnr.unixsocket.SockAddrUnix.create(SockAddrUnix.java:174)
	at jnr.unixsocket.UnixSocketAddress.<init>(UnixSocketAddress.java:53)
	at com.microsoft.azure.sdk.iot.device.hsm.HttpsHsmClient.sendHttpRequestUsingUnixSocket(HttpsHsmClient.java:250)
	at com.microsoft.azure.sdk.iot.device.hsm.HttpsHsmClient.sendRequestBasedOnScheme(HttpsHsmClient.java:226)
	at com.microsoft.azure.sdk.iot.device.hsm.HttpsHsmClient.sign(HttpsHsmClient.java:112)
	at com.microsoft.azure.sdk.iot.device.hsm.HttpHsmSignatureProvider.sign(HttpHsmSignatureProvider.java:85)
	at com.microsoft.azure.sdk.iot.device.hsm.IotHubSasTokenHsmAuthenticationProvider.createNewSasToken(IotHubSasTokenHsmAuthenticationProvider.java:80)
	at com.microsoft.azure.sdk.iot.device.hsm.IotHubSasTokenHsmAuthenticationProvider.create(IotHubSasTokenHsmAuthenticationProvider.java:48)
	at com.microsoft.azure.sdk.iot.device.ModuleClient.createFromEnvironment(ModuleClient.java:235)
	at storemanager.azure.EdgeModuleHandler.createModule(EdgeModuleHandler.java:81)
	at storemanager.azure.EdgeModuleHandler.startModule(EdgeModuleHandler.java:73)
	at storemanager.azure.EdgeModuleHandler.start(EdgeModuleHandler.java:48)
	at storemanager.Application.main(Application.java:14)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:48)
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:87)
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:50)
	at org.springframework.boot.loader.JarLauncher.main(JarLauncher.java:51)

	at com.kenai.jffi.internal.StubLoader.load(StubLoader.java:270)
	at com.kenai.jffi.internal.StubLoader.<clinit>(StubLoader.java:444)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:348)
	at com.kenai.jffi.Init.load(Init.java:68)
	at com.kenai.jffi.Foreign$InstanceHolder.getInstanceHolder(Foreign.java:49)
	at com.kenai.jffi.Foreign$InstanceHolder.<clinit>(Foreign.java:45)
	... 45 more
```

**Comment by timtay-microsoft** [comment]
For the springboot part of this issue, we had another customer run into that issue (detailed [here](https://github.com/Azure/azure-iot-sdk-java/issues/274#issuecomment-455425347)) where the resolution was to use 

```java
<dependency>
<groupId>org.springframework.boot</groupId>
<artifactId>spring-boot-starter-webflux</artifactId>
</dependency>
```

rather than

```java
<dependency>
<groupId>org.springframework.boot</groupId>
<artifactId>spring-boot-starter-web</artifactId>
</dependency>
```

Could you try changing this dependency and then try it out again?

**Comment by hochas** [comment]
We aren't using any webserver, I just thought the information would be helpful when looking for the issue. But I did try to use webflux, and it does not resolve the issue, although the error message got a bit more information:

```
019-02-12 07:03:01.882  INFO 1 --- [           main] storemanager.azure.EdgeModule    : Creating module from environment
[WARN] AnnotationConfigReactiveWebServerApplicationContext - Exception encountered during context initialization - cancelling refresh attempt: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'moduleInitializer' defined in URL [jar:file:/app.jar!/BOOT-INF/classes!/storemanager/azure/ModuleInitializer.class]: Invocation of init method failed; nested exception is java.lang.UnsatisfiedLinkError: could not load FFI provider jnr.ffi.provider.jffi.Provider
2019-02-12 07:03:02.910  INFO 1 --- [           main] s.i.s.a.gatemaster.GateMasterRegistry    : Attempting to close all sockets and device clients
[INFO] ThreadPoolTaskScheduler - Shutting down ExecutorService 'defaultTaskScheduler'
[INFO] ConditionEvaluationReportLoggingListener - 

Error starting ApplicationContext. To display the conditions report re-run your application with 'debug' enabled.
[ERROR] SpringApplication - Application run failed <org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'moduleInitializer' defined in URL [jar:file:/app.jar!/BOOT-INF/classes!/storemanager/azure/ModuleInitializer.class]: Invocation of init method failed; nested exception is java.lang.UnsatisfiedLinkError: could not load FFI provider jnr.ffi.provider.jffi.Provider>org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'moduleInitializer' defined in URL [jar:file:/app.jar!/BOOT-INF/classes!/azure/ModuleInitializer.class]: Invocation of init method failed; nested exception is java.lang.UnsatisfiedLinkError: could not load FFI provider jnr.ffi.provider.jffi.Provider
```
Again, just taking the docker image I pushed to the edge device from the portal and running it as follows works:

``` 
docker run --entrypoint="/opt/jdk/bin/java" image:arm-webflux -jar -Dspring.profiles.active=development app.jar
```
Where the development profile makes the application create the ModuleClient from a provided string instead of using `createFromEnvironment`.

This also works:

```
docker run -e EdgeHubConnectionString="connectionString" image:arm-webflux
```

It also works when we run the application using `iotedgehubdev`, so there is no way  for us to reliably test for this issue other than actually deploying it.

I also tried building the container for amd64 and deploying it to a virtual machine with **and** without any webserver, neither tomcat nor jetty, and it starts in both scenarios.

I also tried completely reinstalling the raspberry from scratch, and the issue still persists.

**Comment by timtay-microsoft** [comment]
From my digging so far, this appears to be an issue with the JFFI dependency that our JNR dependency has. For some context, when on a linux device, this SDK uses unix domain sockets to communicate with an HSM for authentication. Your logs suggest that you are unable to use the unix domain socket library (JNR) to reach that HSM, so the SDK falls back to the connection string that you manually set.  

The version of JNR that this SDK uses is currently 0.19, and judging by [this PR](https://github.com/jnr/jnr-unixsocket/pull/59) to the JNR library, only JNR versions 0.20 and up have support for ARM64 architecture devices. There was a bit more context around this issue explained in [this thread](https://github.com/spotify/docker-client/issues/477) if you are interested.

Can you try cloning this SDK from source, upgrading the JNR dependency, and then try to create from environment (without setting the connection string in the environment variables)? The dependency can be found in the device client POM
```java
<dependency>
    <groupId>com.github.jnr</groupId>
    <artifactId>jnr-unixsocket</artifactId>
    <version>0.19</version>
</dependency>
```

**Comment by hochas** [comment]
Unfortunately it did not resolve the issue. 

I built using this dependency in `iot-sdk-java/device/iot-device-client/pom.xml`:
```
<dependency>
	<groupId>com.github.jnr</groupId>
	<artifactId>jnr-unixsocket</artifactId>
	<version>0.22</version>
</dependency>
```
I pushed to our Azure container registry and deployed the same way as my initial post.
There is a new line this time around:

`Caused by: java.lang.UnsatisfiedLinkError: could not get native definition for type: POINTER`

Here's the updated [stacktrace](https://gist.github.com/hochas/8e97328ea12d2d98baba1e77a94e6dee).

Launching only the docker and starting the module client using `new ModuleClient("connectionString", protocol);` still works.

Is there anything else I can do to help?

**Comment by timtay-microsoft** [comment]
Hmm... I'd suggest you open an issue with the [JFFI library](https://github.com/jnr/jffi) about this issue that you are seeing. As far as I can tell, JNR (which uses JFFI) is not supported on your platform.

Alternatively, you could explore using a different unix domain socket implementation. This dependency only exists in one place, so it should be simple for you to experiment with swapping out to a different implementation.

Within the file HttpsHsmClient.java, there is the following method which is the only place we use unix domain sockets:
```java
private HttpsResponse sendHttpRequestUsingUnixSocket(HttpsRequest httpsRequest, String httpRequestPath, String httpRequestQueryString, String unixSocketAddress) throws IOException
{
    UnixSocketChannel channel = null;
    HttpsResponse response = null;
    try
    {
        //write to socket
        byte[] requestBytes = HttpsRequestResponseSerializer.serializeRequest(httpsRequest, httpRequestPath, httpRequestQueryString, unixSocketAddress);
        UnixSocketAddress address = new UnixSocketAddress(unixSocketAddress);
        channel = UnixSocketChannel.open(address);

        if (httpsRequest.getBody() != null)
        {
            //append http request body to the request bytes
            ByteArrayOutputStream outputStream = new ByteArrayOutputStream( );
            outputStream.write(requestBytes);
            outputStream.write(httpsRequest.getBody());

            channel.write(ByteBuffer.wrap(outputStream.toByteArray()));
        }
        else
        {
            channel.write(ByteBuffer.wrap(requestBytes));
        }

        //read response
        String responseString = readResponseFromChannel(channel);
        response = HttpsRequestResponseSerializer.deserializeResponse(new BufferedReader(new StringReader(responseString)));
    }
    finally
    {
        if (channel != null)
        {
            channel.close();
        }
    }

    return response;
}
```

**Comment by hochas** [comment]
Thank you for all your help, I will try to find another solution when l'm back at work. I will also open a ticket where you suggested. If this issue remains open I will update with what I find.

**Comment by hochas** [comment]
A colleague of mine fetched the related jnr dependencies from master and built them locally and used them in our application, and it works when starting on the architecture mentioned in the original post.

All of the following dependencies were built from master from their corresponding repositories:
```
<dependency>
	<groupId>com.github.jnr</groupId>
	<artifactId>jffi</artifactId>
	<version>1.2.19-SNAPSHOT</version>
</dependency>

<dependency>
	<groupId>com.github.jnr</groupId>
	<artifactId>jnr-ffi</artifactId>
	<version>2.1.10-SNAPSHOT</version>
</dependency>

<dependency>
	<groupId>com.github.jnr</groupId>
	<artifactId>jnr-unixsocket</artifactId>
	<version>0.23-SNAPSHOT</version>
</dependency>
```

**Comment by timtay-microsoft** [comment]
Good to hear. We'll keep an eye out for when those snapshots are officially released, so that this SDK can update its dependencies to fix this issue. Thanks for doing the legwork here!

**Comment by timtay-microsoft** [comment]
JNR recently released jnr-unixsocket-0.23 so I'm going to get our dependency upgraded so that you don't have to compile from source anymore to solve this issue.

**Comment by timtay-microsoft** [comment]
The fix for this issue has been released as of Device Client version 1.17.5, so I'm going to close this issue. Feel free to open a new issue if you still see any problems here.

**Comment by az-iot-builder-01** [comment]
@hochas, thank you for your contribution to our open-sourced project! Please help us improve by filling out this 2-minute [customer satisfaction survey](https://forms.office.com/Pages/ResponsePage.aspx?id=v4j5cvGGr0GRqy180BHbRx8jpG79OudPqY-BojcKjkJUMFFBQjgxNkw5ODJIRjBWTzI0Nk85R0pUQiQlQCN0PWcu)
